File: backend/src/main/java/info/gamewise/lor/videos/service/GetJsonDataService.java
package info.gamewise.lor.videos.service;

import info.gamewise.lor.videos.config.JsonDataProperties;
import info.gamewise.lor.videos.config.LoRCacheConfig;
import info.gamewise.lor.videos.domain.json.Champion;
import info.gamewise.lor.videos.domain.json.Channel;
import info.gamewise.lor.videos.port.out.GetChampionsPort;
import info.gamewise.lor.videos.port.out.GetChannelsPort;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

import static java.util.Collections.singletonList;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.util.CollectionUtils.isEmpty;

@Service
class GetJsonDataService implements GetChannelsPort, GetChampionsPort {

    private static final Logger LOG = LoggerFactory.getLogger(GetJsonDataService.class);
    private static final AtomicReference<List<Channel>> CHANNELS = new AtomicReference<>();
    private static final AtomicReference<List<Champion>> CHAMPIONS = new AtomicReference<>();

    private final JsonDataProperties properties;

    GetJsonDataService(JsonDataProperties properties) {
        this.properties = properties;
    }

    @Override
    @Cacheable(LoRCacheConfig.CHAMPIONS)
    public List<Champion> getChampions() {
        LOG.info("Fetching champions from {}", properties.getChampionsUrl());

        RestTemplate restTemplate = new RestTemplate();
        ResponseEntity<List<Champion>> response = restTemplate.exchange(properties.getChampionsUrl(), GET, httpEntity(), new ParameterizedTypeReference<>() {});

        CHAMPIONS.set(response.getBody());
        return CHAMPIONS.get();
    }

    @Override
    public Optional<Champion> getChampionByName(String champion) {
        if (isEmpty(CHAMPIONS.get())) {
            getChampions();
        }
        return CHAMPIONS.get().stream().filter(c -> champion.equals(c.code())).findFirst();
    }

    @Override
    @Cacheable(LoRCacheConfig.CHANNELS)
    public List<Channel> getChannels() {
        LOG.info("Fetching channels from {}", properties.getChannelsUrl());

        RestTemplate restTemplate = new RestTemplate();
        ResponseEntity<List<Channel>> response = restTemplate.exchange(properties.getChannelsUrl(), GET, httpEntity(), new ParameterizedTypeReference<>() {});

        CHANNELS.set(response.getBody());
        return CHANNELS.get();
    }

    @Override
    public Optional<Channel> getChannelByCode(String channelCode) {
        if (isEmpty(CHANNELS.get())) {
            getChannels();
        }
        return CHANNELS.get().stream().filter(c -> channelCode.equals(c.code())).findFirst();
    }

    private HttpEntity<?> httpEntity() {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(singletonList(APPLICATION_JSON));
        headers.setContentType(APPLICATION_JSON);
        return new HttpEntity<>(headers);
    }
}

File: backend/src/main/kotlin/info/gamewise/lor/videos/service/GetJsonDataService.kt
package info.gamewise.lor.videos.service

import info.gamewise.lor.videos.config.JsonDataProperties
import info.gamewise.lor.videos.config.LoRCacheConfig
import info.gamewise.lor.videos.domain.json.Champion
import info.gamewise.lor.videos.domain.json.Channel
import info.gamewise.lor.videos.port.out.GetChampionsPort
import info.gamewise.lor.videos.port.out.GetChannelsPort
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.Cacheable
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.*
import org.springframework.stereotype.Service
import org.springframework.util.CollectionUtils
import org.springframework.web.client.RestTemplate
import java.util.concurrent.atomic.AtomicReference

@Service
internal open class GetJsonDataService(private val properties: JsonDataProperties) : GetChannelsPort, GetChampionsPort {

    private val log = LoggerFactory.getLogger(GetJsonDataService::class.java)
    private val channels = AtomicReference<List<Channel>>(emptyList())
    private val champions = AtomicReference<List<Champion>>(emptyList())

    @Cacheable(LoRCacheConfig.CHAMPIONS)
    override fun getChampions(): List<Champion> {
        log.info("Fetching champions from {}", properties.championsUrl)
        val restTemplate = RestTemplate()
        val response: ResponseEntity<List<Champion>> = restTemplate.exchange(
            properties.championsUrl,
            HttpMethod.GET,
            httpEntity(),
            object : ParameterizedTypeReference<List<Champion>>() {})
        champions.set(response.body)
        return champions.get()
    }

    override fun getChampionByName(champion: String): Champion? {
        if (champions.get().isEmpty()) {
            getChampions()
        }

        return champions.get().firstOrNull { champion == it.code }
    }

    @Cacheable(LoRCacheConfig.CHANNELS)
    override fun getChannels(): List<Channel> {
        log.info("Fetching channels from {}", properties.channelsUrl)
        val restTemplate = RestTemplate()
        val response: ResponseEntity<List<Channel>> = restTemplate.exchange(
            properties.channelsUrl,
            HttpMethod.GET,
            httpEntity(),
            object : ParameterizedTypeReference<List<Channel>>() {})
        channels.set(response.body)
        return channels.get()
    }

    override fun getChannelByCode(channelCode: String): Channel? {
        if (channels.get().isEmpty()) {
            getChannels()
        }

        return channels.get().firstOrNull { channelCode == it.code }
    }

    private fun httpEntity(): HttpEntity<*> {
        val headers = HttpHeaders()
        headers.accept = listOf(MediaType.APPLICATION_JSON)
        headers.contentType = MediaType.APPLICATION_JSON
        return HttpEntity<Any>(headers)
    }

}